### ISSUE TITLE
Functionality to stream dask arrays to ims files

### ISSUE THREAD

**Issue by m-albert** [tlambert03/imarispy#2]
Hi @tlambert03,

thanks so much for sharing this useful code.

Recently I've been building upon your repo to stream `dask.arrays` to imaris files. As you might know, in `dask`, there's a function `.to_hdf5` which allows you to write the result of a dask computation into an hdf5 array in a chunkwise manner. For creating imaris hdf5 files this is not entirely straight forward, as result chunks should only be computed once to feed into the different resolution levels of a pyramidal file.

Therefore I've slightly adapted `imarispy.np_to_ims.py` to use `dask.array.core.store` in the case of having a dask array as an input. `dask.array.core.store` takes care of optimizing the dask graph underlying the computations. As a result, dask computations that are part of an image processing pipeline can directly stream their results into pyramidal imaris files using your code.

As more and more people are using dask for image processing (of course also in `napari`), the function could be useful to people and probably your repository is the place to go for when saving python arrays into imaris files! Anyways, just wanted to let you know that the above mentioned code is available [here](https://github.com/m-albert/imarispy/blob/dask_array_to_ims/imarispy/imaris.py). If you're interested, I can create a PR to extend your code.

Cheers,
Marvin

**Comment by m-albert** [comment]
Here's an example code working with https://github.com/m-albert/imarispy/blob/dask_array_to_ims/imarispy/imaris.py

```python
import numpy as np
import imarispy

# create image
im = np.random.randint(0, 100, (20, 30, 40), dtype=np.uint16)

# numpy array to ims
import os
fn = '/tmp/myfile_np.ims'
if os.path.exists(fn):
    os.remove(fn)
imarispy.np_to_ims(im, fn)

# create dask array
import dask.array as da
im_da = da.from_array(im)

# dask array to ims
fn = '/tmp/myfile_da.ims'
if os.path.exists(fn):
    os.remove(fn)
imarispy.np_to_ims(im_da, fn)
```

**Comment by tlambert03** [comment]
Thanks Marvin, that sounds great.  I haven't been using this repo much recently, but I'd love to have that integrated here if you'd like to submit a PR.  (actually, I've been working more on [`napari`](https://github.com/napari/napari) recently, and I can see this sort of thing also being useful for interoperability there).

I will also point out that @nvladimus took some of the original ideas here and made much better library ([npy2bdv](https://github.com/nvladimus/npy2bdv))... however, I don't believe he maintained the `.ims` support there. given the similarity between the formats, it'd be awesome if this could all live under one roof, and you're probably more familiar with the imaris side of things than I am at this point.  @nvladimus, would you consider having a module in npy2bdv that handled `.ims` mapping?  or would you prefer to keep that bdv?

**Comment by m-albert** [comment]
> (actually, I've been working more on napari recently, and I can see this sort of thing also being useful for interoperability there).

I agree, definitely this could also be useful in the context of `napari`! The extension to dask arrays, but in the first place also your original `np_tp_ims` would be very useful for the community (and of course the bdv support by @nvladimus).

I'm wondering where it would make most sense to add. Maybe in `napari/utils/io.py` it'd be a bit out of context? Also I agree that it would be nice to have ims and bdv living under one roof. Curious to know what @nvladimus thinks!

**Comment by tlambert03** [comment]
We're not putting any file-type specific stuff in napari, so this would need to be implemented as a [napari plugin](https://napari.org/docs/plugins/index.html).  For true BDV support of course there's a lot more metadata to support, see https://github.com/napari/napari/issues/1308 for discussion. But I do think it would be a nice (perhaps independent) venture to simply create a plugin that converts an ims file to a dask array so that napari can read native imaris files!

If you're interested in helping to get Imaris file support in napari, I'd be happy to help you write that plugin.  (It could ultimately live here or in the other repo... doesn't matter to me)

**Comment by m-albert** [comment]
> We're not putting any file-type specific stuff in napari, so this would need to be implemented as a napari plugin.

That makes sense of course! Once I have some more time I'll try to write a `napari` plugin. Thanks for all the useful info and links!

**Comment by nvladimus** [comment]
Hi, guys! This sounds very interesting. I'd love to extend the `npy2bdv` library to a more general one that also handles `.ims` format.  I'll look into this.

**Comment by nvladimus** [comment]
Edited my earlier response after reading the imarispy code again. Imaris format looks clear now. I guess I was sleep deprived when writing my earlier questions :-)
